### PR TITLE
Fixes #4117: Add DigitalOcean Tag support

### LINKED ISSUE #4117: digitalocean tag support
##### ISSUE TYPE
- Feature Idea
##### COMPONENT NAME

digital_ocean
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file = /Users/vivo/workspaces/me/ansible-digitalocean/ansible.cfg
  configured module search path = ['/Users/vivo/workspaces/me/ansible-modules-core']
```
##### CONFIGURATION

N/A
##### OS / ENVIRONMENT

N/A
##### SUMMARY

DigitalOcean now supports tagging of Droplets.
Ansible should offer support for using tags

E.g. to ensure that a tag exists:

```
  digital_ocean: >
    state=present
    command=tag
    name="foo"
    api_token="{{ lookup('env','DO_API_TOKEN') }}"
```

I am already working on adding the support. Just wanted to create an issue to avoid duplication of effort

### PR DESCRIPTION
**Pull request by zyegfryed**
##### ISSUE TYPE

New Module Pull Request
##### COMPONENT NAME

cloud/digitalocean
##### ANSIBLE VERSION

```
ansible 2.1.0.0
  config file =
  configured module search path = Default w/o overrides
```
##### SUMMARY

[Fixes #4117] Add tag support to DigitalOcean module. 
Rely on `requests` libraray instead of `dopy`.

### CODE REVIEW

File: cloud/digital_ocean/digital_ocean_tag.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+
+
+def core(module):
+    try:
+        api_token = module.params['api_token'] or \
+                os.environ['DO_API_TOKEN'] or os.environ['DO_API_KEY']
+    except KeyError as e:
+        module.fail_json(msg='Unable to load %s' % e.message)
+
+    state = module.params['state']
```
Comment (kontrafiktion) [inline cloud/digital_ocean/digital_ocean_tag.py:125]: What happens if state or name are not present. Is the error message human understandable?

File: cloud/digital_ocean/digital_ocean_tag.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    name = module.params['name']
+    resource_id = module.params['resource_id']
```
Comment (kontrafiktion) [inline cloud/digital_ocean/digital_ocean_tag.py:127]: I think the resource _name_ should be supported as well (resource_id and resource_name should be mutually exclusive)

File: cloud/digital_ocean/digital_ocean_tag.py (hunk 1 above)
Comment (zyegfryed) [inline cloud/digital_ocean/digital_ocean_tag.py:125]: This is handled in the `main()` function when we instantiate `AnsibleModule`. `name` parameter is required and `state` is optional and defaults to `present`.

**Comment by kontrafiktion** [comment]
+1 to creating a new module and not hacking it into the "digital_ocean" 
+1 for not using dopy

Instead of using the requests library (which is great), we should consider using ansible.module_utils.urls (especially the "open_url" method) to reduce dependencies

**Comment by kontrafiktion** [comment]
Travis already remarked:

ERROR: No RETURN documentation provided
ERROR: requests import found, should use ansible.module_utils.urls instead
ERROR: GPLv3 license header not found

File: cloud/digital_ocean/digital_ocean_tag.py (hunk 2 above)
Comment (zyegfryed) [inline cloud/digital_ocean/digital_ocean_tag.py:127]: The `resource_id` relates to a DO resource. AFAIK there's no `resource_name` in DO API [Tag a Resource](https://developers.digitalocean.com/documentation/v2/#tag-a-resource) entry.

File: cloud/digital_ocean/digital_ocean_tag.py (hunk 1 above)
Comment (kontrafiktion) [inline cloud/digital_ocean/digital_ocean_tag.py:125]: ok

File: cloud/digital_ocean/digital_ocean_tag.py (hunk 2 above)
Comment (kontrafiktion) [inline cloud/digital_ocean/digital_ocean_tag.py:127]: ok, then we might add that later on (querying for droplets with the given name and extracting the ID ourselves)

**Comment by gregdek** [comment]
Thanks @zyegfryed for this new module. When this module receives 'shipit' comments from two community members and any 'needs_revision' comments have been resolved, we will mark for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by zyegfryed** [comment]
I've added the RETURN documentation and the GPLv3 license header.
Regarding the `requests` import, unless someone give me some documentation pointer and/or code sample on how to properly interrogate a JSON API with `ansible.module_utils.urls`, I'm gonna leave it as is. Feel free to hack it.

**Comment by gregdek** [comment]
Thanks @zyegfryed for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by kontrafiktion** [comment]
created a new pull request #4209 with “requests” replaced

## 

France is Bacon!

**Comment by zyegfryed** [comment]
@kontrafiktion You rock! Thanks!

**Comment by zyegfryed** [comment]
Closing that PR since PR#4218 supersedes it

**Comment by kontrafiktion** [comment]
@zyegfryed our PR has been merged. thx for letting me help

**Comment by zyegfryed** [comment]
👍

**Pull request closed without merging**